### PR TITLE
fix concurrent-map-write of FakeOS.Create

### PR DESCRIPTION
**Pull request by silenceshell**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:

This PR adds a `sync.RWMutex` to `FakeOS` to protect map `Files` from concurrent writes.
We met this bug when creating quite a lot of pods on nodes created by kubemark, here is the stack:

```
fatal error: concurrent map writes
I0207 09:26:08.307373       1 kuberuntime_manager.go:428] No sandbox for pod "test-base-mid-hollow-node-0-0-6ff6bff99f-8hmqk_press-hollow-node-0-0(9c3204a3-0817-4291-9b33-df10bcb583c0)" can be found. Need to start a new one

goroutine 304 [running]:
runtime.throw(0x4201de6, 0x15)
	/usr/local/go/src/runtime/panic.go:1116 +0x72 fp=0xc0013d9e08 sp=0xc0013d9dd8 pc=0x43b332
runtime.mapassign_faststr(0x3947600, 0xc000a01530, 0xc001b56bd0, 0x65, 0x65)
	/usr/local/go/src/runtime/map_faststr.go:211 +0x3f1 fp=0xc0013d9e70 sp=0xc0013d9e08 pc=0x417ed1
k8s.io/kubernetes/pkg/kubelet/container/testing.(*FakeOS).Create(0xc0003c5a40, 0xc001b56bd0, 0x65, 0xc001b56bd0, 0x65, 0xc001a05f60)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/container/testing/os.go:87 +0x5a fp=0xc0013d9ea8 sp=0xc0013d9e70 pc=0x3320a7a
k8s.io/kubernetes/pkg/kubelet/kuberuntime.(*kubeGenericRuntimeManager).makeMounts(0xc00037a500, 0xc000caa0b0, 0xc0005fa000, 0x0, 0xc0013024e0, 0x0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kuberuntime/kuberuntime_container.go:353 +0x362 fp=0xc0013d9ff0 sp=0xc0013d9ea8 pc=0x3205ec2
k8s.io/kubernetes/pkg/kubelet/kuberuntime.(*kubeGenericRuntimeManager).generateContainerConfig(0xc00037a500, 0xc0005fa000, 0xc000758c00, 0x0, 0xc0007f2390, 0xb, 0xc000be8050, 0x48, 0xc00108ed40, 0x1, ...)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kuberuntime/kuberuntime_container.go:282 +0x599 fp=0xc0013da168 sp=0xc0013d9ff0 pc=0x3205159
k8s.io/kubernetes/pkg/kubelet/kuberuntime.(*kubeGenericRuntimeManager).startContainer(0xc00037a500, 0xc0008da140, 0x43, 0xc00022d730, 0xc000c37310, 0xc000758c00, 0xc000b24400, 0x6734ee8, 0x0, 0x0, ...)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kuberuntime/kuberuntime_container.go:166 +0x5dd fp=0xc0013da4f8 sp=0xc0013da168 pc=0x320331d
k8s.io/kubernetes/pkg/kubelet/kuberuntime.(*kubeGenericRuntimeManager).SyncPod.func1(0x41d7145, 0x9, 0xc000c37310, 0xc00022d700, 0x0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kuberuntime/kuberuntime_manager.go:810 +0x887 fp=0xc0013da788 sp=0xc0013da4f8 pc=0x3222107
k8s.io/kubernetes/pkg/kubelet/kuberuntime.(*kubeGenericRuntimeManager).SyncPod(0xc00037a500, 0xc000758c00, 0xc000b24400, 0x6734ee8, 0x0, 0x0, 0xc00017f700, 0xc000fa0020, 0x4, 0x4, ...)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kuberuntime/kuberuntime_manager.go:849 +0xbe5 fp=0xc0013db190 sp=0xc0013da788 pc=0x3216145
k8s.io/kubernetes/pkg/kubelet.(*Kubelet).syncPod(0xc000662000, 0x0, 0xc000758c00, 0x2, 0xc000b24400, 0x0, 0xc0007de0c0, 0xc0005ace40)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1717 +0xfb7 fp=0xc0013dbd30 sp=0xc0013db190 pc=0x32ec357
k8s.io/kubernetes/pkg/kubelet.(*Kubelet).syncPod-fm(0x0, 0xc000758c00, 0x2, 0xc000b24400, 0x0, 0x0, 0xc000b24400)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1503 +0x48 fp=0xc0013dbd80 sp=0xc0013dbd30 pc=0x331db88
k8s.io/kubernetes/pkg/kubelet.(*podWorkers).managePodLoop.func1(0xc0007d7ec0, 0xc0001fe5b0, 0xc0013dbea8, 0x43dfdc, 0xc0010e9e50)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/pod_workers.go:175 +0x24a fp=0xc0013dbe28 sp=0xc0013dbd80 pc=0x331c06a
k8s.io/kubernetes/pkg/kubelet.(*podWorkers).managePodLoop(0xc0001fe5b0, 0xc00101e540)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/pod_workers.go:184 +0x145 fp=0xc0013dbf80 sp=0xc0013dbe28 pc=0x33138a5
k8s.io/kubernetes/pkg/kubelet.(*podWorkers).UpdatePod.func1(0xc0001fe5b0, 0xc00101e540)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/pod_workers.go:222 +0x6d fp=0xc0013dbfd0 sp=0xc0013dbf80 pc=0x331c16d
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1374 +0x1 fp=0xc0013dbfd8 sp=0xc0013dbfd0 pc=0x473581
created by k8s.io/kubernetes/pkg/kubelet.(*podWorkers).UpdatePod
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/pod_workers.go:220 +0x3c7
```

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/kubelet/container/testing/os.go [hunk 1]
```diff
@@ -77,7 +79,9 @@ func (f *FakeOS) Create(path string) (*os.File, error) {
 	if f.Files == nil {
 		f.Files = make(map[string][]*os.FileInfo)
 	}
+	f.FilesLock.Lock()
 	f.Files[path] = []*os.FileInfo{}
+	f.FilesLock.Unlock()
```
Comment (wzshiming) [inline pkg/kubelet/container/testing/os.go:84]:
```suggestion
	f.FilesLock.Lock()
	defer f.FilesLock.Unlock()
	if f.Files == nil {
		f.Files = make(map[string][]*os.FileInfo)
	}
	f.Files[path] = []*os.FileInfo{}
```

File: pkg/kubelet/container/testing/os.go [hunk 2]
```diff
@@ -113,11 +117,13 @@ func (f *FakeOS) ReadDir(dirname string) ([]os.DirEntry, error) {
 func (f *FakeOS) Glob(pattern string) ([]string, error) {
 	if f.GlobFn != nil {
 		var res []string
+		f.FilesLock.RLock()
```
Comment (wzshiming) [inline pkg/kubelet/container/testing/os.go:120]:
```suggestion
		f.FilesLock.RLock()
		defer f.FilesLock.RUnlock()
```

**Review by wzshiming — CHANGES_REQUESTED** [review]
A small change using `defer`

**Comment by wzshiming** [comment]
https://github.com/kubernetes/kubernetes/blob/15b63c380ee6d8ea7443deaa0928f8380b6985a7/pkg/kubelet/kuberuntime/kuberuntime_container_test.go#L82-L83

I've looked at the code and this is the only place it is called how does it trigger a concurrent map write?

File: pkg/kubelet/container/testing/os.go [hunk 3]
```diff
@@ -77,12 +79,14 @@ func (f *FakeOS) Create(path string) (*os.File, error) {
 	if f.Files == nil {
 		f.Files = make(map[string][]*os.FileInfo)
 	}
+	f.FilesLock.Lock()
+	defer f.FilesLock.Unlock()
```
Comment (wzshiming) [inline pkg/kubelet/container/testing/os.go:83]: Move to the top of the function.

**Review by wzshiming — CHANGES_REQUESTED** [review]

**Review by wzshiming — APPROVED** [review]
/lgtm

**Comment by silenceshell** [comment]
@wzshiming Would you help to merge this PR? Thanks.

**Comment by wzshiming** [comment]
> @wzshiming Would you help to merge this PR? Thanks.

Sorry, this requires an [approver](https://github.com/kubernetes/kubernetes/blob/05f451b58f0d5c54b534acf472fb530fe016ddaf/OWNERS_ALIASES#L219-L226) replay `/approve`

**Comment by silenceshell** [comment]
@SergeyKanzhelev @bart0sh

**Comment by wzshiming** [comment]
/cc @thockin
/cc @mrunalp

Could you have time to take a look?
This patch is a simple fix for a problem in kubemark that caused concurrent map writes due to fast Pod creation.

**Comment by silenceshell** [comment]
@liggitt PTAL

**Comment by derekwaynecarr** [comment]
Verified the lock is protecting where the map is modified and read.

/approve
/lgtm

**Pull request merged**